optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,418 @@
+            distributions.UniformDistribution,
+            distributions.LogUniformDistribution,
+            distributions.IntUniformDistribution,
+            distributions.IntLogUniformDistribution,
+            distributions.DiscreteUniformDistribution,
+        )
+        rescaled_samples_list = []
+        for param_name in param_names:
+            if isinstance(self._search_space[param_name], continuous_param_types):
+                high = self._high[param_name]
+                low = self._low[param_name]
+                assert high is not None
+                assert low is not None
+                samples = multivariate_samples[param_name]
+                samples = (samples - low) / (high - low)
+                rescaled_samples_list.append(samples)
+            else:  # Categorical parameters are ignored.
+                continue
+
+        if len(rescaled_samples_list) == 0:
+            return None
+        elif len(rescaled_samples_list[0]) == 0:
+            return np.ones(1, dtype=float)
+
+        rescaled_samples = np.array(rescaled_samples_list).T
+
+        # compute distance matrix of samples
+        distances = np.linalg.norm(
+            rescaled_samples[:, None, :] - rescaled_samples[None, :, :], axis=2
+        )
+        distances += np.diag(np.ones(len(samples)) * np.infty)
+
+        return np.min(distances, axis=1)
+
+    def _calculate_categorical_params(self, samples: np.ndarray, param_name: str) -> np.ndarray:
+
+        samples = samples.astype(int)
+        distribution = self._search_space[param_name]
+        assert isinstance(distribution, distributions.CategoricalDistribution)
+        choices = distribution.choices
+        consider_prior = self._parameters.consider_prior
+        prior_weights = self._parameters.prior_weight
+        if consider_prior:
+            weights = prior_weights / samples.size * np.ones((samples.size + 1, len(choices)))
+            weights[np.arange(samples.size), samples] += 1
+        else:
+            weights = prior_weights / samples.size * np.ones((samples.size, len(choices)))
+            weights[np.arange(samples.size), samples] += 1
+        weights /= weights.sum(axis=1, keepdims=True)
+        return weights
+
+    def _calculate_parzen_est_params(
+        self, samples: np.ndarray, param_name: str
+    ) -> Tuple[np.ndarray, np.ndarray]:
+
+        samples_size = samples.size
+        consider_prior = self._parameters.consider_prior
+        consider_magic_clip = self._parameters.consider_magic_clip
+        sigmas0 = self._sigmas0
+        low = self._low[param_name]
+        high = self._high[param_name]
+        assert sigmas0 is not None
+        assert low is not None
+        assert high is not None
+
+        if samples_size == 0:
+            consider_prior = True
+
+        if consider_prior:
+
+            prior_mu = 0.5 * (low + high)
+            prior_sigma = 1.0 * (high - low)
+
+            mus = np.zeros(samples_size + 1)
```

Comment:
This is also a detail but we only have to allocate memory here and not iterate over (fill) it.
```suggestion
            mus = np.empty(samples_size + 1)
```

Reply:
Thanks for nice revision!